src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -138,7 +207,44 @@ private void ReflectionSetMemberValue(object obj, object memberValue, MemberInfo
                 else
                 {
                     object property = propInfo.GetValue(obj);
-                    MethodInfo addMethod = property.GetType().GetMethod("Add");
+                    Type propertyType = property.GetType();
+                    MethodInfo addMethod = propertyType.GetMethod("Add");
+
+                    if (addMethod.GetParameters().Length != 1)
+                    {
+                        var enumerator = ((IEnumerable)memberValue).GetEnumerator();
+
+                        if (!enumerator.MoveNext() || enumerator.Current == null)
+                            return;
+
+                        Type itemType = enumerator.Current.GetType();
+                        Type genericCollectionType = Globals.TypeOfICollectionGeneric.MakeGenericType(itemType);
+                        if (genericCollectionType.IsAssignableFrom(propertyType))
+                        {
+                            addMethod = genericCollectionType.GetMethod("Add");
+                        }
+                        else if (Globals.TypeOfIDictionary.IsAssignableFrom(propertyType) && Globals.TypeOfIDictionary.IsAssignableFrom(memberValue.GetType()))
+                        {
+                            var sourceDict = (IDictionary)memberValue;
+                            var targetDict = (IDictionary)property;
+                            foreach (var key in sourceDict.Keys)
+                            {
+                                var value = sourceDict[key];
+                                targetDict.Add(key, value);
+                            }
+
+                            return;
+                        }
+                        //else if (Globals.TypeOfIList.IsAssignableFrom(propertyType))
+                        //{
+                        //     addMethod = Globals.TypeOfIList.GetMethod("Add");
+                        //}
+                        else
+                        {
+                            throw new InvalidOperationException(string.Format("Cannot set the member: {0} of type: {1}", memberInfo.Name, obj.GetType().Name));
```

Comment:
Should this be a localized string?  Who will see it and what will they do when they see it?  Example: should we try to say why we could not set the member?